pallets/gated-marketplace/src/functions.rs

```diff
@@ -150,9 +151,328 @@ impl<T: Config> Pallet<T> {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+
+        //insert in OffersInfo
+        // ensure the offer_id doesn't exist
+        ensure!(!<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferAlreadyExists);
+        <OffersInfo<T>>::insert(offer_id, offer_data);
+
+        //Insert in OffersByMarketplace
+        <OffersByMarketplace<T>>::try_mutate(marketplace_id, |offers| {
+            offers.try_push(offer_id)
+        }).map_err(|_| Error::<T>::OfferStorageError)?;
+    
+        Self::deposit_event(Event::OfferStored(collection_id, item_id));
+        Ok(())
+    }
+
+    pub fn do_take_sell_offer(buyer: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId,) -> DispatchResult {
+        //This extrisicn is called by the user who wants to buy the item
+        //ensure the collection & owner exists
+        let owner_item = pallet_uniques::Pallet::<T>::owner(collection_id, item_id).ok_or(Error::<T>::OwnerNotFound)?;
+
+        //ensure owner is not the same as the buyer
+        ensure!(owner_item != buyer.clone(), Error::<T>::CannotTakeOffer); 
+
+        //ensure the selected item has a valid offer_id in OffersInfo
+        ensure!(<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferNotFound);
+
+        //ensure the offer_id exists in OffersByItem
+        Self::does_exist_offer_id_for_this_item(collection_id, item_id, offer_id)?;
+
+        //TODO: ensure the offer is not expired
+
+        //ensure the offer is open and available
+        ensure!(Self::is_offer_status(offer_id, OfferStatus::Open), Error::<T>::OfferIsNotAvailable);
+       
+        //Transfer balance to the owner of the item
+        let item_price =  Self::get_offer_price(offer_id).map_err(|_| Error::<T>::OfferNotFound)?;
+        let total_user_balance = T::LocalCurrency::total_balance(&buyer);
+        ensure!(total_user_balance >= item_price, Error::<T>::NotEnoughBalance);
+        //Transfer the balance
+        T::LocalCurrency::transfer(&buyer, &owner_item, item_price, KeepAlive)?;
+
+        //Use uniques transfer function to transfer the item to the buyer
+        pallet_uniques::Pallet::<T>::do_transfer(collection_id, item_id, buyer.clone(), |_, _|{
+            Ok(())
+        })?;
+
+        //update offer status from all marketplaces
+        Self::update_offers_status(buyer.clone(), collection_id, item_id, marketplace_id)?;
+
+        //remove all the offers associated with the item
+        Self::delete_all_offers_for_this_item(collection_id, item_id )?;
+        
+        //TODO: add the offer_id from this offer to the buyer's history
+
+        Self::deposit_event(Event::OfferWasAccepted(offer_id, buyer));
+        Ok(())
+    }
+
+    pub fn do_take_buy_offer(authority: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId,) -> DispatchResult {
```

Comment:
That's right, all current signature functions corresponds to the version 4. When I updated the offers flow to version 5 I moved some parameters into OfferData<T> structure, (Marketplace_id, collection_id & item_id). I'll change it.

Reply:
Update: I haven't removed the marketplace_id because we need to know that the marketplace_id where the extrinsic is being called is the same as the one in the selected offer.

Reply:
Done.